force_wfmanager/left_side_pane/variable_names_registry.py

```diff
@@ -50,50 +57,68 @@ def __init__(self, workflow, *args, **kwargs):
 
     @on_trait_change(
         'workflow.mco.parameters.name,'
-        'workflow.execution_layers.data_sources.output_slot_info.name')
+        'workflow.execution_layers.data_sources.output_slot_info.name,'
+        'workflow.mco.parameters.type,'
+    )
     def update_available_variables_stack(self):
         stack = []
 
         # At the first layer, the available variables are the MCO parameters
         if self.workflow.mco is None:
             stack.append([])
         else:
-            stack.append([
-                p.name
-                for p in self.workflow.mco.parameters
-                if len(p.name) != 0
-            ])
+            stack.append([{'name': p.name, 'type': p.type}
+                         for p in self.workflow.mco.parameters
+                         if len(p.name) != 0
+                          ])
 
         for layer in self.workflow.execution_layers:
             stack_entry_for_layer = []
             for ds in layer.data_sources:
-                stack_entry_for_layer.extend([
-                    info.name
-                    for info in ds.output_slot_info
-                    if info.name != ''])
+                ds_names = [info.name for info in ds.output_slot_info]
+                ds_output_slots = ds.factory.create_data_source().slots(ds)[1]
```

Comment:
this code is in the hands of the plugin implementer, so it could throw exceptions. uhm....

Reply:
I think it may not be as easy to have a VariableNames object if we want types as well. It works at the minute because we can get all the information from the Workflow, but we need data which can only be found in the ModelView(s). 
So I think the VariableNames functionality will have to be implemented across the ModelViews themselves as well as in VariableNames, due to the fact that mvs are recreated often. 
Also I realised that as well as relying on plugin code, calling get/create_data_source() does not account for slot changes, so it's not a good solution unfortunately!